使用 SQLite 做 Database

Li-Pro opened the pull request:
支援 db.py 裡所計的幾個 schema，將他們包成一個主 database，並支援 CRUD (create, query/query_value/read, update, delete) 的操作。以下是一些細節：
- 支援 create (格式是固定的，但是自動欄位 (auto_increment, default, nullable 等) 可以不必填寫
- 支援以 `dict` 為基礎，比較相符資料的查詢 (read/query_value)
- 支援更進階的 SQL 查詢/操作 (query/update/delete)，因為細節複雜且之後可能會換 DBMS，
  所以使用第三方 `pypika` 做 SQL 的 Builder。採用直接的 criteria，語法用起來還蠻方便的 (參考 tests)。
- 支援多筆(bulk)更新。在 Python 迴圈處理資料實在太慢了。
- 密碼儲存使用 `bcrypt` 來做 hashing。
- test/ 裡面包含很多測試(& 範例)，包含 database, table, utility 都有測試 (coverage 100% (X))
- 因為有裝套件所以我簡單弄了 Pipenv，不過也可以用其他的代替。

期間我做了一些 refactor，之後要新增新的 table 也會比較方便。  
我有把 fake_db_init 和 main 裡面使用到 db 的地方改寫，應該可以直接用了 (沒有直接從 API 測試過)。

Li-Pro committed:
Initialize python dev environment

Li-Pro committed:
feat: Create database table types with native database connection

Li-Pro committed:
feat: Add dep PyPika as database query builder

Li-Pro committed:
feat: Implement CRUD with pypika query

Li-Pro committed:
feat: Implement user table query with equal filter

Li-Pro committed:
fix: Fix field typo and display command debug logs

Li-Pro committed:
feat: Support bulk dictionary-based query

Li-Pro committed:
feat: Add tests for user data table

Li-Pro committed:
feat: Support symptom database table

Li-Pro committed:
refactor: Move general database operation to base table

Li-Pro committed:
doc: Document DatabaseNative

Li-Pro committed:
feat: Support user database schema v1.0

Li-Pro committed:
feat: Add untested bulletin table and clinic table

Li-Pro committed:
Merge branch 'sql' into simple_db_add_sqlite

Include database schema in db source code (unused). Vim's
gitignore is included in incoming branch so it's replaced.

Li-Pro committed:
feat: Implement fake_db master database interface

Li-Pro committed:
feat: Add (failed) fake_db testcase

Li-Pro committed:
feat: Support fake_db's favor of database operation

Li-Pro committed:
perf: Operate database in bulk, in fake_db

Li-Pro committed:
feat: Save password in bcrypt hash instead of plaintext

Li-Pro commented on main.py:
```diff
@@ -24,15 +24,18 @@
-})
+from db import MasterDatabase
+
+db = MasterDatabase(".local/test.db")
+guest = db.create(
+    "users",
+    {
+        "username": "guest",
+        "password": "no_password",
+        "permission": 0,
+        "auth_method": "no_password",
+    },
+)
```
用法大概沒變，就是改成叫 MasterDatabase，然後要指定資料庫位置而已。  
架構是 MasterDatabase 管理各個 table (*Database，都是 BaseTableDatabase)，使用 DatabaseNative 操作 SQLite。

ZoneTwelve commented on Pipfile.lock:
@Li-Pro Could you double-check if this file is actually required to be present on the GitHub repos?

Li-Pro replied:
It's the same mechanic as `package.json` and `package-lock.json`. It records package locked hash for package dependency reproduction, so it's fine (and better) included.

Ref: [Pipfile.lock](https://stackoverflow.com/questions/58743814/pipenv-lock-file-depolyment), [package-lock.json](https://medium.com/@InspireTech/should-i-add-package-lock-json-to-git-why-is-it-important-to-commit-package-lock-json-into-git-e8273a8aad00#:~:text=By%20committing%20the%20package%2Dlock,any%20given%20point%20in%20time.)

Li-Pro replied:
It's fine to use `requirements.txt` though. I picked it just for simple package environment reproduction. Would you prefer that I changed it to `requirements.txt`?

ZoneTwelve replied:
Thank you for taking the time to reply.

Li-Pro committed:
Merge branch 'simple_db' into simple_db_add_sqlite

ZoneTwelve commented:
@Li-Pro Could you please take a look at the files with conflicts and resolve the discrepancies?

Li-Pro committed:
feat: Update test according to new schema

Li-Pro committed:
feat!: Support class auto-creation based on property map

Breaking change: Database property format is modified to
suit furture extension.

Li-Pro committed:
feat!: Move primary key to column attribute

Breaking change: Change of database property.

Li-Pro committed:
feat: Support array with read conversion

Li-Pro committed:
feat: Implement parameterized data creation

Li-Pro committed:
feat: Support custom type criterion converter

Li-Pro committed:
feat: Suppport data operation custom type converter

Li-Pro commented:
@ZoneTwelve 
The conflicts were from the schema update, so I updated the database tables according to it.  
It's worth noting that:
- updating the schema might require a database migration in future, though we don't really have important data now
- SQLite does NOT support a raw "array" type, and few implementations would support dynamic-length array, so I implemented custom array serialize/de-serialize for this case for now. This would also aid in future custom type integration.

By the way, I have updated the format of schema specification (DB_PROPERTY) so that it suits database table attributes better.

Li-Pro committed:
feat!: Return key-value map for entry creation

Li-Pro commented on db.py:
```diff
@@ -1,87 +1,509 @@
+        },
     },
-    'bulletins': {
-        'bulletin_id': {'dtype': 'int', 'auto_increment': True},
-        'class': {'dtype': 'str'},
-        'user_id': {'dtype': 'int'},
-        'title': {'dtype': 'str'},
-        'content': {'dtype': 'str'},
-        'update_at': {'dtype': 'str'},
-        'create_at': {'dtype': 'str'},
+    "bulletins": {
+        "columns": {
+            "bulletin_id": {"dtype": "int", "primary": True, "auto_increment": True},
+            "class": {"dtype": "str"},
+            "user_id": {"dtype": "int"},
+            "title": {"dtype": "str"},
+            "content": {"dtype": "str"},
+            "update_at": {"dtype": "str"},
+            "create_at": {"dtype": "str"},
+        },
+    },
+    "clinics": {
+        "columns": {
+            "clinic_id": {"dtype": "int", "primary": True, "auto_increment": True},
+            "title": {"dtype": "str"},
+            "address": {"dtype": "str"},
+            "tel": {"dtype": "str"},
+            "tags": {"dtype": "array"},
+            "owner_id": {"dtype": "int"},
+        },
     },
-    'clinics': {
-        'clinic_id': {'dtype': 'int', 'auto_increment': True},
-        'title': {'dtype': 'str'},
-        'address': {'dtype': 'str'},
-        'tel': {'dtype': 'str'},
-        'tags': {'dtype': 'array'},
-        'owner_id': {'dtype': 'int'},
-    }
 }
```
This is the updated schema specification format (the format we parse from). The schema itself is unchanged.

I gave up creating the table using the exported SQL format (the SQL commands exported from the database diagram), since it was not SQLite-compatible nor a fit for the features we needed.
